src/scanspec/service.py

```diff
@@ -1,161 +1,302 @@
-                Points(self._chunk.midpoints[axis]),
-                Points(self._chunk.upper[axis]),
-            )
-            for axis in self._chunk.axes()
-        ]
-
-    @resolver
-    def smallest_abs_step(self) -> float:
-        """The smallest step between midpoints across ALL axes in the scan."""
-        return _calc_smallest_step(list(self._chunk.midpoints.values()))
+class MidpointsResponse(GeneratedPointsResponse):
+    """Midpoints of a generated scan."""
 
+    midpoints: Mapping[str, Points] = Field(
+        description="The midpoints of scan frames for each axis"
+    )
 
-def _calc_smallest_step(points: List[np.ndarray]) -> float:
-    # Calc abs diffs of all axes
-    absolute_diffs = [abs_diffs(axis_midpoints) for axis_midpoints in points]
-    # Return the smallest value (Aka. smallest step)
-    return np.amin(np.linalg.norm(absolute_diffs, axis=0))
 
+@dataclass
+class BoundsResponse(GeneratedPointsResponse):
+    """Bounds of a generated scan."""
 
-def abs_diffs(array: np.ndarray) -> np.ndarray:
-    """Calculates the absolute differences between adjacent elements in the array.
+    lower: Mapping[str, Points] = Field(
+        description="Lower bounds of scan frames if different from midpoints"
+    )
+    upper: Mapping[str, Points] = Field(
+        description="Upper bounds of scan frames if different from midpoints"
+    )
+
+
+@dataclass
+class GapResponse:
+    """Presence of gaps in a generated scan."""
+
+    gap: List[bool] = Field(
+        description="Boolean array indicating if there is a gap between each frame"
+    )
+
+
+@dataclass
+class SmallestStepResponse:
+    """Information about the smallest steps between points in a spec."""
+
+    absolute: float = Field(
+        description="Absolute smallest distance between two points on a single axis"
+    )
+    per_axis: Mapping[str, float] = Field(
+        description="Smallest distance between two points on each axis"
+    )
+
+
+#
+# API Routes
+#
+
+_EXAMPLE_SPEC = Line("y", 0.0, 10.0, 3) * Line("x", 0.0, 10.0, 4)
+_EXAMPLE_POINTS_REQUEST = PointsRequest(
+    _EXAMPLE_SPEC, max_frames=1024, format=PointsFormat.FLOAT_LIST
+)
+
+
+@app.post("/valid", response_model=ValidResponse)
+def valid(
+    spec: Spec = Body(..., example=_EXAMPLE_SPEC)
+) -> Union[ValidResponse, JSONResponse]:
+    """Validate wether a ScanSpec can produce a viable scan.
 
     Args:
-        array: A 1xN array of numerical values
+        spec: The scanspec to validate
 
     Returns:
-        A newly constucted array of absolute differences
+        ValidResponse: A canonical version of the spec if it is valid.
+            An error otherwise.
     """
-    # [array[1] - array[0], array[2] - array[1], ...]
-    adjacent_diffs = array[1:] - array[:-1]
-    return np.absolute(adjacent_diffs)
+    valid_spec = Spec.deserialize(spec.serialize())
+    return ValidResponse(spec, valid_spec)
+
+
+@app.post("/midpoints", response_model=MidpointsResponse)
```

Comment:
What does GDA actually want from this service? If just midpoints then this is fine, if midpoints, bounds, smallest step, then we calculate the spec 3 times which is quite wasteful

Reply:
I suspect just midpoints initially, bounds and smallest step may come later. I was going to suggest adding a caching mechanism when required, since I doubt it will want the extra information for every request.

Reply:
I guess that is the REST way, caching and multiple requests, rather than multiple arguments changing what comes out in the response